File: creator-app/src/bot/keyboard.ts
import { TabListEntry, BotCommand, CallStatus } from '../types';

export function generateShortId(tabId: string): number {
  let hash = 0;
  for (let i = 0; i < tabId.length; i++) {
    hash = ((hash << 5) - hash) + tabId.charCodeAt(i);
    hash = hash & hash;
  }
  return Math.abs(hash) % 10000;
}

export function padShortId(id: number): string {
  return ('000' + id).slice(-4);
}

export function findTabByShortId(tabsList: TabListEntry[], shortId: string): TabListEntry | null {
  const target = parseInt(shortId, 10);
  for (const tab of tabsList) {
    if (generateShortId(tab.id) === target) return tab;
  }
  return null;
}

export function createMainKeyboard() {
  return {
    one_time: false,
    buttons: [
      [
        { action: { type: 'text', label: '🟦 VK', payload: JSON.stringify({ cmd: BotCommand.VK, mode: 'headless' }) } },
        { action: { type: 'text', label: '🟥 Telemost', payload: JSON.stringify({ cmd: BotCommand.TM, mode: 'headless' }) } },
      ],
      [
        { action: { type: 'text', label: '🟪 WBStream', payload: JSON.stringify({ cmd: BotCommand.WB, mode: 'headless' }) } },
        { action: { type: 'text', label: '🟩 DION', payload: JSON.stringify({ cmd: BotCommand.Dion, mode: 'headless' }) } },
      ],
      [
        { action: { type: 'text', label: '🔗 Join by link', payload: JSON.stringify({ cmd: BotCommand.JoinPrompt }) } },
      ],
      [
        { action: { type: 'text', label: '📋 Active Tabs', payload: JSON.stringify({ cmd: BotCommand.List }) } },
      ],
      [
        { action: { type: 'text', label: 'ℹ️ Buttons below: old webview mode', payload: JSON.stringify({ cmd: BotCommand.Noop }) } },
      ],
      [
        { action: { type: 'text', label: '🟦 DC (legacy)', payload: JSON.stringify({ cmd: BotCommand.VK, mode: 'dc' }) } },
        { action: { type: 'text', label: '🟦 Video (legacy)', payload: JSON.stringify({ cmd: BotCommand.VK, mode: 'video' }) } },
        { action: { type: 'text', label: '🟥 Video (legacy)', payload: JSON.stringify({ cmd: BotCommand.TM, mode: 'video' }) } },
      ],
    ],
  };
}

export function createWaitingKeyboard() {
  return {
    one_time: false,
    buttons: [
      [
        { action: { type: 'text', label: '◀️ Back', payload: JSON.stringify({ cmd: BotCommand.Menu }) } },
      ],
    ],
  };
}

export function createListKeyboard(tabsList: TabListEntry[]) {
  const buttons = tabsList.map((entry) => {
    const shortId = padShortId(generateShortId(entry.id));
    const prefix = entry.isBot ? 'bot' : 'user';
    const status = entry.callStatus === CallStatus.Active ? '🟢' : '⚪';
    return [
      {
        action: {
          type: 'text',
          label: `${prefix} ${entry.platform} ${entry.mode} ${shortId} ${status}`,
          payload: JSON.stringify({ cmd: BotCommand.Close, id: entry.id }),
        },
      },
    ];
  });
  buttons.push([
    { action: { type: 'text', label: '◀️ Back', payload: JSON.stringify({ cmd: BotCommand.Menu }) } },
  ]);
  return { one_time: false, buttons };
}

File: headless/vk-bot/keyboard.go
package main

import "encoding/json"

type kbAction struct {
	Type    string `json:"type"`
	Label   string `json:"label"`
	Payload string `json:"payload"`
}

type kbButton struct {
	Action kbAction `json:"action"`
}

type keyboard struct {
	OneTime bool         `json:"one_time"`
	Buttons [][]kbButton `json:"buttons"`
}

func platformSquare(platform string) string {
	switch platform {
	case "vk":
		return "🟦"
	case "tm", "telemost":
		return "🟥"
	case "wb", "wbstream":
		return "🟪"
	case "dion":
		return "🟩"
	case "bitrix":
		return "🟧"
	}
	return ""
}

func mainMenuKeyboard() string {
	kb := keyboard{
		OneTime: false,
		Buttons: [][]kbButton{
			{
				{Action: kbAction{Type: "text", Label: "🟦 VK", Payload: `{"cmd":"vk"}`}},
				{Action: kbAction{Type: "text", Label: "🟥 Telemost", Payload: `{"cmd":"tm"}`}},
			},
			{
				{Action: kbAction{Type: "text", Label: "🟪 WBStream", Payload: `{"cmd":"wb"}`}},
				{Action: kbAction{Type: "text", Label: "🟩 DION", Payload: `{"cmd":"dion"}`}},
			},
			{
				{Action: kbAction{Type: "text", Label: "🟧 Bitrix", Payload: `{"cmd":"bitrix"}`}},
			},
			{
				{Action: kbAction{Type: "text", Label: "🔗 Join by link", Payload: `{"cmd":"join-prompt"}`}},
			},
			{
				{Action: kbAction{Type: "text", Label: "📋 Active sessions", Payload: `{"cmd":"list"}`}},
			},
		},
	}
	data, _ := json.Marshal(kb)
	return string(data)
}

func waitingKeyboard() string {
	kb := keyboard{
		OneTime: false,
		Buttons: [][]kbButton{
			{
				{Action: kbAction{Type: "text", Label: "◀️ Back", Payload: `{"cmd":"menu"}`}},
			},
		},
	}
	data, _ := json.Marshal(kb)
	return string(data)
}

func sessionsKeyboard(sessions []*session) string {
	rows := make([][]kbButton, 0, len(sessions)+1)
	for _, s := range sessions {
		payload, _ := json.Marshal(map[string]string{"cmd": "close", "id": s.id})
		label := platformSquare(s.platform) + " " + s.platform + " " + s.id + " 🟢"
		rows = append(rows, []kbButton{
			{Action: kbAction{Type: "text", Label: label, Payload: string(payload)}},
		})
	}
	rows = append(rows, []kbButton{
		{Action: kbAction{Type: "text", Label: "◀️ Back", Payload: `{"cmd":"menu"}`}},
	})
	kb := keyboard{OneTime: false, Buttons: rows}
	data, _ := json.Marshal(kb)
	return string(data)
}
